pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,166 @@
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+const (
+	DeleteQueueName = "pod_delete"
+	PodIPIndex      = "PodIP"
+	DelayTime       = time.Minute * 5
+)
+
+type PodStore struct {
+	curPod       cache.Indexer
+	prevPod      cache.Indexer
+	podsToDelete workqueue.DelayingInterface
+}
+
+// newPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func newPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPod:       podInformer.GetIndexer(),
+		prevPod:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			DeleteFunc: s.onPodDelete,
+		})
+	s.curPod.AddIndexers(cache.Indexers{PodIPIndex: podIPIndexFunc})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return newPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing event delete")
+			return
+		}
+	}
+	if pod.GetDeletionTimestamp() == nil {
+		timeNow := time.Now()
+		pod.DeletionTimestamp = &metav1.Time{Time: timeNow}
+	}
+	klog.V(4).InfoS("pod timestamp", "pod name", pod.ObjectMeta.Name, "creationTime", pod.GetCreationTimestamp().String(), "deletionTime", pod.GetDeletionTimestamp().String())
+	s.prevPod.Add(pod)
+	s.podsToDelete.AddAfter(pod, DelayTime)
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool) {
+	// Query current Pod first
+	curPods, _ := s.curPod.ByIndex(PodIPIndex, ip)
+	if len(curPods) != 0 {
+		pod := curPods[0].(*corev1.Pod)
+		if pod.GetCreationTimestamp().Time.Before(startTime) {
+			klog.V(4).InfoS("Get pod info from current store", "podName", pod.Name, "podNamespace", pod.Namespace)
+			return pod, true
+		}
+	}
+	prePods, _ := s.prevPod.ByIndex(PodIPIndex, ip)
+	if len(prePods) == 0 {
+		return nil, false
+	}
+	for _, prePod := range prePods {
+		pod := prePod.(*corev1.Pod)
+		if pod.GetCreationTimestamp().Time.Before(startTime) && startTime.Before(pod.GetDeletionTimestamp().Time) {
```

Comment:
I don't feel great about relying on `CreationTimestamp` and `DeletionTimestamp` for this, for a couple of reasons.
The main reason is that they are set by the K8s apiserver (unless I am mistaken), using a clock that may be skewed compared to the local one. On the other hand, "startTime" comes from conntrack data, and is based on the local clock.

I feel like it would be better to call `clock.Now()` when we get the `ADDED` & `DELETED` events for the Pod, and use that as the creation and deletion timestamps. This is more in line with the recommendation in https://github.com/kubernetes/kubernetes/issues/87039#issuecomment-1457208334.

@tnqn what do you think?

Reply:
But if can't modify the object in `Added` or `Deleted` event, how can we have an accurate `clock.Now()`?
Or we need to cache this information somewhere in the Pod Store and use pod.uuid to find the correct time later?

Reply:
Agree with @antoninbas. `DeletionTimestamp` is actually the timestamp calcualted based on the time kube-apiserver receives the DELETE request plus the Pod's `TerminationGracePeriodSeconds`, not when the Pod is actually deleted.
Using local clock makes sense to me. There are many ways you can achieve it without modifying the Pod object, like creating a struct which contains a pointer to the original Pod, a local creationTimestamp, a local deletionTimestamp, or maintain separate maps of PodUID/PodName -> local CreationTimestamp/DeletionTimestamp. The former would look like:
```
type PodWithLocalClock struct {
	*Pod
	// The local clock when the Pod is added
	CreationTimestamp time.Time
        // The local clock when the Pod is actually deleted from kube-apiserver. Nil if it's not deleted yet.
	DeletionTimestamp *time.Time
}
```

Reply:
I've updated it to record the creationTimestamp/DeletionTimestamp when we have Add/Delete events by using the local clock and a map.